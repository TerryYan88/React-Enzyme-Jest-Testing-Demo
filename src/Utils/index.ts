/* eslint-disable react/forbid-foreign-prop-types */
import {ShallowWrapper} from "enzyme";
import checkPropTypes from "check-prop-types";
import { applyMiddleware,createStore } from "redux";
import { rootReducer } from "../redux/reducers/rootReducer";


export const testStore = (initialState={}) =>{
    const createStoreWithMiddleware = applyMiddleware()(createStore)
    return createStoreWithMiddleware(rootReducer,initialState);
}


export const findByTestAttr = (component: ShallowWrapper,attr:string)=>{
    const wrapper = component.find(`[data-test="${attr}"]`);
    return wrapper;
}

export const checkProps = (component:any,expectedProps:{})=>{
    const propsErr = checkPropTypes(component.propTypes,expectedProps,'props',component.name);
    return propsErr;
}


