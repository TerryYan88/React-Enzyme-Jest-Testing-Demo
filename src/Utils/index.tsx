
import {ShallowWrapper} from "enzyme";


export const findByTestAttr = (component: ShallowWrapper,attr:string)=>{
    const wrapper = component.find(`[data-test="${attr}"]`);
    return wrapper;
}