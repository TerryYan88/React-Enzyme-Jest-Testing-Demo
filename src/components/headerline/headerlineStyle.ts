import styled from "styled-components";


export const HeaderLineContainer = styled.div`
    

`