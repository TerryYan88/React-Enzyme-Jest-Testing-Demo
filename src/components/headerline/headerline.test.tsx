import React from 'react';
import renderer from "react-test-renderer";
import { shallow, ShallowWrapper } from 'enzyme';
import HeaderLine from './headerline';
import { findByTestAttr } from "../../Utils";
import { HeaderLineContainer } from "./headerlineStyle";
import {checkProps} from "../../Utils"
interface ItempArr {
    fName: string,
    lName: string,
    email: string,
    age: number,
    onlineStatus: boolean,
}
interface HeaderLineProps {
    header: string;
    desc: string;
    tempArr: Array<ItempArr>;
}

const setUp = (props: HeaderLineProps) => {
    const component = shallow(<HeaderLine {...props!} />)
    return component;
}

describe('HeaderLine Component', () => {
    it("snapshot",()=>{
        const props: HeaderLineProps = {
            header: "Test Header",
            desc: "Test Dest",
            tempArr: [{
                fName: "Test fName",
                lName: "Test lName",
                email: "Test email",
                age: 0,
                onlineStatus: true,
            }]
        }
        const component = (<HeaderLine {...props}/>);
        const tree = renderer.create(component).toJSON();
        expect(tree).toMatchSnapshot();
    })
    describe('have props', () => {
        let component: ShallowWrapper;
        beforeEach(() => {
            const props: HeaderLineProps = {
                header: "Test Header",
                desc: "Test Dest",
                tempArr: [{
                    fName: "Test fName",
                    lName: "Test lName",
                    email: "Test email",
                    age: 0,
                    onlineStatus: true,
                }]
            }
            component = setUp(props);
        })

        it("should render without errors", () => {
            const container = component.find(HeaderLineContainer);
            expect(container.length).toBe(1);
            const headLineComponent = findByTestAttr(component, "HeadlineComponent");
            expect(headLineComponent.length).toBe(1);
        })
        it("should render a H1", () => {
            const h1 = findByTestAttr(component, "header");
            expect(h1.length).toBe(1);
        })
        it("should render a desc", () => {
            const desc = findByTestAttr(component, "desc");
            expect(desc.length).toBe(1);
        })

    })
    describe("checking prop-types",()=>{
        it("should not throw a warning",()=>{
            const expectedProps = {
                header:"Test header",
                desc:"Test desc",
                tempArr: [{
                    fName: "Test fName",
                    lName: "Test lName",
                    email: "Test@email.com",
                    age: 23,
                    onlineStatus: false,
                }]
            }
            const propsErr = checkProps(HeaderLine,expectedProps)
            expect(propsErr).toBeUndefined();
        })
    })
    // describe("have not props", () => {
    //     let component: ShallowWrapper;
    //     beforeEach(() => {
    //         component = setUp();
    //     })
    //     it("should render headerline component", () => {
    //         const container = component.find(HeaderLineContainer);
    //         expect(container.length).toBe(1);
    //         const headLineComponent = findByTestAttr(component, "HeadlineComponent");
    //         expect(headLineComponent.length).toBe(1);
    //     })
    //     it("should not render H1", () => {
    //         const h1 = findByTestAttr(component, "header");
    //         expect(h1.length).toBe(0);
    //     })
    //     it("should not render a desc", () => {
    //         const desc = findByTestAttr(component, "desc");
    //         expect(desc.length).toBe(0);
    //     })
    // })
})