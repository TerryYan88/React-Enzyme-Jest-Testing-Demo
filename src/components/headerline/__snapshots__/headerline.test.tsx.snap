// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderLine Component snapshot 1`] = `
<div
  className=""
  data-test="HeadlineComponent"
>
  <h1
    data-test="header"
  >
    Test Header
  </h1>
  <p
    data-test="desc"
  >
    Test Dest
  </p>
</div>
`;
